### PR TITLE
Conform to naming conventions for attributes

### PR DESCRIPTION
**Pull request by romanlutz**
Related to #256  (but doesn't fully fix it since Postprocessing needs to be addressed as well)

Adjust the attributes to naming conventions, specifically for an attribute named `attr`:
- everything set during `__init__` becomes `attr`
- everything set during `fit` becomes `attr_` unless we want to keep it hidden (then `_attr`)

Since we have example notebooks where we use some of these, e.g. `_predictors`, we should expose them properly.

Additionally, this cleanup includes:
- providing `run_lp_step` as an argument in ExponentiatedGradient. Right now it's a hard-coded constant which isn't very useful for people who want to actually use it.
- renaming two arguments for ExponentiatedGradient, namely 
  - `T` to `max_iterations` to make it more obvious
  - `eta_mul` to `learning_rate` to imply the purpose

### CODE REVIEW

File: fairlearn/reductions/_exponentiated_gradient/_lagrangian.py [hunk 1]
```diff
@@ -48,17 +49,16 @@ def __init__(self, X, sensitive_features, y, estimator, constraints, eps, B, opt
...
-        self.errors = pd.Series(dtype="float64")
-        self.gammas = pd.DataFrame()
-        self.lambdas = pd.DataFrame()
-        self.n = self.X.shape[0]
-        self.n_oracle_calls = 0
-        self.n_oracle_calls_dummy_returned = 0
-        self.oracle_execution_times = []
-        self.last_linprog_n_hs = 0
-        self.last_linprog_result = None
+        self.hs_ = pd.Series(dtype="float64")
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/_lagrangian.py:52]: I don't think that we need to be using this convention in classes that are not estimators. I'd rather revert these.

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 2]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
         violation is at most :code:`2*(eps+best_gap)`
     :type eps: float
 
-    :param T: Maximum number of iterations
-    :type T: int
+    :param max_iterations: Maximum number of iterations
+    :type max_iterations: int
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:40]: if we are renaming this, let's follow [sklearn](https://scikit-learn.org/stable/glossary.html#parameters): `max_iter`

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 3]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
...
     :type nu: float
 
-    :param eta_mul: Initial setting of the learning rate
-    :type eta_mul: float
+    :param learning_rate: Initial setting of the learning rate
+    :type learning_rate: float
+
+    :param run_lp_step: if True each step of exponentiated gradient is followed by the saddle
+        point optimization over the convex hull of classifiers returned so far; default True
+    :type run_lp_step: bool
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:51]: maybe `run_lp_step` -> `linprog_step`?

Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:51]: `linprog` seems easier to read.

File: fairlearn/reductions/_exponentiated_gradient/_lagrangian.py (hunk 1 above)
Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/_lagrangian.py:52]: Sounds good!

File: CHANGES.md [hunk 4]
```diff
@@ -25,6 +29,8 @@
 * Fix warning due to changing default `dtype` when creating an empty
   `pandas.Series`.
 * Enable `GridSearch` for more than two sensitive features values.
+* Change `ExponentiatedGradient` signature by renaming argument `T` to
+  `max_iterations`, `eta_mul` to `learning_rate`, and by adding `run_lp_step`.
 * Add new disparity constraints for reductions methods as moments in
```
Comment (MiroDudik) [inline CHANGES.md:34]: remember to update the names here according to corrections

Comment (romanlutz) [reply CHANGES.md:34]: ... and it's also in the wrong release because it didn't make 0.4.6 :-)

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 5]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
...
+        self.max_iterations = max_iterations
+        self.nu = nu
+        self.learning_rate = learning_rate
+        self.run_lp_step = run_lp_step
+
+        self.best_gap_ = None
+        self.predictors_ = None
+        self.weights_ = None
+        self.last_t_ = None
+        self.best_t_ = None
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:68]: `last_iter_` and `best_iter_` to match `max_iter`?

Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:68]: Good point! I like that

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 6]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
...
+        self.n_oracle_calls_ = 0
+        self.n_oracle_calls_dummy_returned_ = 0
+        self.oracle_execution_times_ = None
+        self.lambda_vecs_ = pd.DataFrame()
+        self.lambda_vecs_LP_ = pd.DataFrame()
+        self.lambda_vecs_lagrangian_ = pd.DataFrame()
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:75]: Actually, I think that we will need to tweak these names:
* `lambda_vecs_` -> `lambda_vecs_EG_` [the dimensionality of EG and LP variants is `n_iter_`]
* `lambda_vecs_lagrangian_` ->`lambda_vecs_` [this is because it matches the dimension of `predictors_` and `weights_`]

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 5 above)
Comment (MiroDudik) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py]: also... i wonder whether we don't want to replace `last_iter_` by `n_iter_` which would be equal to `last_iter_+1`

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 7]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
...
 
     :param nu: Convergence threshold for the duality gap, corresponding to a
         conservative automatic setting based on the statistical uncertainty in measuring
         classification error
     :type nu: float
 
-    :param eta_mul: Initial setting of the learning rate
-    :type eta_mul: float
+    :param learning_rate: Initial setting of the learning rate
+    :type learning_rate: float
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:47]: i'm not sure how i feel about renaming this one... I know that this better matches the current description, but the actual initial learning rate is `eta_mul / B`. Plus, `sklearn` is also using `eta0` in this context: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html

Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:47]: okay, if this is commonly used we can do `eta0`. I just didn't like it to be `mul` because that kind of conflicts with the "initial setting" description. I expected it to be multiplied with `eta` (learning rate). `eta0` makes intuitively sense.

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 6 above)
Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:75]: Done!

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 8]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
...
 
-    :param eta_mul: Initial setting of the learning rate
-    :type eta_mul: float
+    :param eta0: Initial setting of the learning rate
+    :type eta0: float
+
+    :param run_linprog_step: if True each step of exponentiated gradient is followed by the saddle
+        point optimization over the convex hull of classifiers returned so far; default True
+    :type run_linprog_step: bool
     """
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:52]: i'm being a bit nitpicky. but can we just do `linprog_step` rather than `run_linprog_step`?

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py [hunk 9]
```diff
@@ -35,37 +35,43 @@ class ExponentiatedGradient(BaseEstimator, MetaEstimatorMixin):
...
+                 eta0=2.0, run_linprog_step=True):  # noqa: D103
+        self.estimator = estimator
+        self.constraints = constraints
+        self.eps = eps
+        self.max_iter = max_iter
+        self.nu = nu
+        self.eta0 = eta0
+        self.run_linprog_step = run_linprog_step
+
+        self.best_gap_ = None
```
Comment (MiroDudik) [inline fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:64]: These initializations should not be here, but at the very beginning of fit (only if we need them). See: https://github.com/scikit-learn/scikit-learn/blob/fd237278e/sklearn/utils/validation.py#L956

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 8 above)
Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:52]: That doesn't soudn like a bool, though.

File: fairlearn/reductions/_grid_search/grid_search.py [hunk 10]
```diff
@@ -90,12 +90,12 @@ def __init__(self,
...
-        self._lambda_vecs = pd.DataFrame()
-        self._objectives = []
-        self._gammas = pd.DataFrame()
-        self._oracle_execution_times = []
+        self.best_grid_index_ = None
+        self.predictors_ = []
+        self.lambda_vecs_ = pd.DataFrame()
+        self.objectives_ = []
+        self.gammas_ = pd.DataFrame()
+        self.oracle_execution_times_ = []
```
Comment (MiroDudik) [inline fairlearn/reductions/_grid_search/grid_search.py:98]: There's a lot of stuff in this `__init__` that should go into `fit`. But we can hold off for a future PR?

Comment (romanlutz) [reply fairlearn/reductions/_grid_search/grid_search.py:98]: Just did that locally. :-) Thanks!

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 9 above)
Comment (romanlutz) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:64]: Most of them are not even needed. We just didn't like the practice of suddenly defining things outside of `__init__` so far but that's simply how Python folks do that...

Comment (MiroDudik) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:64]: We can possibly hold off with this for a future PR--your call.

File: fairlearn/reductions/_grid_search/grid_search.py [hunk 11]
```diff
@@ -90,12 +90,12 @@ def __init__(self,
         self.grid_offset = grid_offset
         self.grid = grid
 
-        self._best_grid_index = None
-        self._predictors = []
-        self._lambda_vecs = pd.DataFrame()
-        self._objectives = []
-        self._gammas = pd.DataFrame()
-        self._oracle_execution_times = []
+        self.best_grid_index_ = None
```
Comment (MiroDudik) [inline fairlearn/reductions/_grid_search/grid_search.py]: since we're in business of renaming things. i think i might prefer `best_index_` or even `best_idx_` over `best_grid_index` (is it an index of the best grid?)

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 7 above)
Comment (MiroDudik) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py]: this is fine...

File: fairlearn/reductions/_grid_search/grid_search.py (hunk 10 above)
Comment (MiroDudik) [reply fairlearn/reductions/_grid_search/grid_search.py]: this is reasonable. there's also the validation code and a conversion using `float`, but we can tackle that in a future PR.

File: fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py (hunk 8 above)
Comment (MiroDudik) [reply fairlearn/reductions/_exponentiated_gradient/exponentiated_gradient.py:52]: sure

File: notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb [hunk 12]
```diff
@@ -227,7 +227,7 @@
     "lgb_params = {\n",
     "    'objective' : 'binary',\n",
     "    'metric' : 'auc',\n",
-    "    'learning_rate': 0.03,\n",
+    "    'eta0': 0.03,\n",
```
Comment (MiroDudik) [inline notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb:230]: this is a bug. revert back to `learning_rate`.

Comment (romanlutz) [reply notebooks/Binary Classification with the UCI Credit-card Default Dataset.ipynb:230]: Yes, found that last week :-) Already changed locally but I ran into some more issues that I'm fixing now.

File: fairlearn/reductions/_grid_search/grid_search.py (hunk 11 above)
Comment (romanlutz) [reply fairlearn/reductions/_grid_search/grid_search.py]: renamed to `best_idx_`

File: fairlearn/reductions/_grid_search/grid_search.py (hunk 10 above)
Comment (romanlutz) [reply fairlearn/reductions/_grid_search/grid_search.py]: The validation code needs to be addressed here because it requires initialized members which we don't have anymore now :-)
